src/node_report.cc

```diff
@@ -0,0 +1,1115 @@
+  pid_t pid = getpid();
+#endif
+
+  // Determine the required report filename. In order of priority:
+  //   1) supplied on API 2) configured on startup 3) default generated
+  char filename[NR_MAXNAME + 1] = "";
+  if (name != nullptr && strlen(name) > 0) {
+    // Filename was specified as API parameter, use that
+    snprintf(filename, sizeof(filename), "%s", name);
+  } else if (strlen(report_filename) > 0) {
+    // File name was supplied via start-up option, use that
+    snprintf(filename, sizeof(filename), "%s", report_filename);
+  } else {
+    // Construct the report filename, with timestamp, pid and sequence number
+    snprintf(filename, sizeof(filename), "%s", "node-report");
+    seq++;
+#ifdef _WIN32
+    snprintf(&filename[strlen(filename)], sizeof(filename) - strlen(filename),
+             ".%4d%02d%02d.%02d%02d%02d.%d.%03d.txt",
+             tm_struct.wYear, tm_struct.wMonth, tm_struct.wDay,
+             tm_struct.wHour, tm_struct.wMinute, tm_struct.wSecond,
+             pid, seq);
+#else  // UNIX, OSX
+    snprintf(&filename[strlen(filename)], sizeof(filename) - strlen(filename),
+             ".%4d%02d%02d.%02d%02d%02d.%d.%03d.txt",
+             tm_struct.tm_year+1900, tm_struct.tm_mon+1, tm_struct.tm_mday,
+             tm_struct.tm_hour, tm_struct.tm_min, tm_struct.tm_sec,
+             pid, seq);
+#endif
+  }
+
+  // Open the report file stream for writing. Supports stdout/err,
+  // user-specified or (default) generated name
+  std::ofstream outfile;
+  std::ostream* outstream = &std::cout;
+  if (!strncmp(filename, "stdout", sizeof("stdout") - 1)) {
+    outstream = &std::cout;
+  } else if (!strncmp(filename, "stderr", sizeof("stderr") - 1)) {
+    outstream = &std::cerr;
+  } else {
+    // Regular file. Append filename to directory path if one was specified
+    if (strlen(report_directory) > 0) {
+      char pathname[NR_MAXPATH + NR_MAXNAME + 1] = "";
+#ifdef _WIN32
+      snprintf(pathname, sizeof(pathname),
+               "%s%s%s", report_directory, "\\", filename);
+#else
+      snprintf(pathname, sizeof(pathname),
+               "%s%s%s", report_directory, "/", filename);
+#endif
+      outfile.open(pathname, std::ios::out);
+    } else {
+      outfile.open(filename, std::ios::out);
+    }
+    // Check for errors on the file open
+    if (!outfile.is_open()) {
+      if (strlen(report_directory) > 0) {
+        std::cerr << "\nFailed to open Node.js report file: "
+                  << filename << " directory: " << report_directory
+                  << " (errno: " << errno << ")\n";
+      } else {
+        std::cerr << "\nFailed to open Node.js report file: "
+                  << filename << " (errno: " << errno << ")\n";
+      }
+      return;
+    } else {
+      std::cerr << "\nWriting Node.js report to file: " << filename << "\n";
+    }
+  }
+
+  // Pass our stream about by reference, not by copying it.
+  std::ostream &out = outfile.is_open() ? outfile : *outstream;
```

Comment:
style nit: left-leaning references and pointers

Reply:
done